scanner/keppel/main.go

```diff
@@ -66,121 +67,134 @@ func main() {
+	var wg sync.WaitGroup
+
+	// Start worker goroutines
+	for i := 0; i < maxWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for comp := range componentCh {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					processComponent(ctx, comp, keppelScanner, keppelProcessor)
+				}
+			}
+		}()
+	}
 
-	for _, account := range accounts {
-		go HandleAccount(fqdn, account, keppelScanner, keppelProcessor, &wg)
+	// Feed components to workers
+	for _, comp := range components {
+		componentCh <- comp
 	}
+	close(componentCh)
 
+	// Wait for all workers to finish
 	wg.Wait()
 }
 
-func HandleAccount(fqdn string, account models.Account, keppelScanner *scanner.Scanner, keppelProcessor *processor.Processor, wg *sync.WaitGroup) error {
-	defer wg.Done()
-	repositories, err := keppelScanner.ListRepositories(account.Name)
+func processComponent(ctx context.Context, comp *client.Component, keppelScanner *scanner.Scanner, keppelProcessor *processor.Processor) {
+	imageInfo, err := keppelScanner.ExtractImageInfo(comp.Name)
 	if err != nil {
-		log.WithFields(log.Fields{
-			"account:": account.Name,
-		}).WithError(err).Error("Error during listing ProcessRepository")
-		return err
+		log.WithError(err).Error("Couldn't extract image information from component name")
+		return
 	}
 
-	for _, repository := range repositories {
-		HandleRepository(fqdn, account, repository, keppelScanner, keppelProcessor)
+	log.Infof("Processing component: %s", comp.Name)
+	compVersions, err := keppelProcessor.GetComponentVersions(comp.Id)
+	if err != nil {
+		log.WithError(err).Errorf("couldn't fetch component versions for componentId: %s", comp.Id)
+		return
 	}
 
-	return nil
-}
-
-func HandleRepository(fqdn string, account models.Account, repository models.Repository, keppelScanner *scanner.Scanner, keppelProcessor *processor.Processor) {
-	component, err := keppelProcessor.ProcessRepository(fqdn, account, repository)
-	if err != nil {
-		log.WithFields(log.Fields{
-			"account:":   account.Name,
-			"repository": repository.Name,
-		}).WithError(err).Error("Error during ProcessRepository")
-		component, err = keppelProcessor.GetComponent(fmt.Sprintf("%s/%s/%s", fqdn, account.Name, repository.Name))
-		if err != nil {
-			log.WithFields(log.Fields{
-				"account:":   account.Name,
-				"repository": repository.Name,
-			}).WithError(err).Error("Error during GetComponent")
-		}
+	for _, cv := range compVersions {
+		HandleImageManifests(ctx, comp.Id, cv.Id, imageInfo.Account, imageInfo.FullRepository(), keppelScanner, keppelProcessor)
 	}
+}
 
-	manifests, err := keppelScanner.ListManifests(account.Name, repository.Name)
+func HandleImageManifests(
+	ctx context.Context,
+	componentId string,
+	componentVersionId string,
+	account string,
+	repository string,
+	keppelScanner *scanner.Scanner,
+	keppelProcessor *processor.Processor,
+) {
+
+	log.Info("Handling manifest")
+	manifests, err := keppelScanner.ListManifests(account, repository)
```

Comment:
We are still listing all manifests? We got the concrete manifest from "component + component version" no?

Reply:
Yes, we can directly fetch the manifest

Reply:
Implemented in [33b65ed](https://github.com/cloudoperators/heureka/pull/309/commits/33b65edb68388ba945d2038d6ce25802756bfb32)